crates/http/src/req.rs

```diff
@@ -1,36 +1,48 @@
 // SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
 // SPDX-License-Identifier: Apache-2.0
 
-use std::{convert::Infallible, future::Future};
+use crate::res::IntoResponse;
 
-use async_trait::async_trait;
-use http_types::{Body, Mime, Request, Response, StatusCode};
+use std::{fmt::Display, future::Future, str::FromStr};
 
-use crate::res::IntoResponse;
+use async_trait::async_trait;
+use http_types::{Body, Error, Mime, Request, Response, Result, StatusCode};
 
 #[async_trait]
 pub trait FromRequest: Sized {
-    type Error: IntoResponse;
-
-    async fn from_request(req: &mut Request) -> Result<Self, Self::Error>;
+    async fn from_request(req: &mut Request) -> Result<Self>;
 }
 
 #[async_trait]
 impl FromRequest for Body {
-    type Error = Infallible;
-
-    async fn from_request(req: &mut Request) -> Result<Self, Self::Error> {
+    async fn from_request(req: &mut Request) -> Result<Self> {
         Ok(req.take_body())
     }
 }
 
+pub fn parse_header<T>(req: &Request, name: &str) -> Result<T>
+where
+    T: FromStr,
+    T::Err: Display,
+{
+    req.header(name)
+        .ok_or_else(|| {
+            Error::from_str(StatusCode::BadRequest, format!("`{}` header not set", name))
+        })?
+        .as_str()
+        .parse()
+        .map_err(|e| {
+            Error::from_str(
+                StatusCode::BadRequest,
+                format!("Could not parse `{}` header value: {}", name, e),
+            )
+        })
+}
```

Comment:
What about this?

```rust
trait FromHeader: FromStr {
    const HEADER: &'static str;
}

impl<T: FromHeader> FromRequest for T { ... }
```

Reply:
I was expecting to get a suggestion about this, went with the simplest/quickest approach in the PR.

We could use this approach, but with the suggested implementation we could only provide a blanket `FromRequest` implementation for `FromHeader` and no other trait until specialization feature is supported by the language.
If that's a limitation we are fine with, I will refactor as suggested

Reply:
@rvolosatovs Let's try it as an experiment. If we hate it, we can undo it.

Reply:
Done in https://github.com/profianinc/drawbridge/pull/36/commits/bbe987c6241877c9f8155f9d393d1ab5f6253d48 (separate commit in case we want to revert in the future)